redact/index.ts

```diff
@@ -0,0 +1,191 @@
+  entity: ArcjetSensitiveInfoType | Custom,
+) {
+  // TODO: Make these the same as analyze.js
+  if (entity === "email") {
+    return { tag: "email" as const };
+  }
+
+  if (entity === "phone-number") {
+    return { tag: "phone-number" as const };
+  }
+
+  if (entity === "ip-address") {
+    return { tag: "ip-address" as const };
+  }
+
+  if (entity === "credit-card-number") {
+    return { tag: "credit-card-number" as const };
+  }
+
+  return {
+    tag: "custom" as const,
+    val: entity,
+  };
+}
+
+function wasmEntitiesToString(entity: SensitiveInfoEntity): string {
+  if (entity.tag === "email") {
+    return "email";
+  }
+
+  if (entity.tag === "ip-address") {
+    return "ip-address";
+  }
+
+  if (entity.tag === "credit-card-number") {
+    return "credit-card-number";
+  }
+
+  if (entity.tag === "phone-number") {
+    return "phone-number";
+  }
+
+  return entity.val;
+}
+
+function performReplacementInText(
+  text: string,
+  replacement: string,
+  start: number,
+  end: number,
+): string {
+  return text.substring(0, start) + replacement + text.substring(end);
+}
+
+function noOpFn(..._args: any[]): any {}
+
+interface RedactedSensitiveInfoEntity {
+  original: string;
+  redacted: string;
+  start: number;
+  end: number;
+  identifiedType: string;
+}
+
+async function callRedactWasm<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
+  CustomEntities extends string,
+>(
+  candidate: string,
+  options: RedactOptions<Detect, CustomEntities>,
+): Promise<RedactedSensitiveInfoEntity[]> {
+  let convertedDetect = noOpFn;
+  if (typeof options.detect !== "undefined") {
+    const detect = options.detect;
+    convertedDetect = (tokens: string[]) => {
+      return detect(tokens)
+        .filter((e) => typeof e !== "undefined")
+        .map((e) => userEntitiesToWasm(e));
+    };
+  }
+
+  let convertedRedact = noOpFn;
+  if (typeof options.replace !== "undefined") {
+    const replace = options.replace;
+    convertedRedact = (identifiedType: SensitiveInfoEntity) => {
+      return replace(wasmEntitiesToString(identifiedType));
+    };
+  }
+
+  const wasm = await initializeWasm(convertedDetect, convertedRedact);
+
+  const skipCustomDetect = options.detect === undefined;
+  const skipCustomRedact = options.redact === undefined;
+
+  const config = {
+    entities: options.redact.map(userEntitiesToWasm),
+    contextWindowSize: options.contextWindowSize,
+    skipCustomDetect,
+    skipCustomRedact,
+  };
+
+  if (typeof wasm !== "undefined") {
+    return wasm.redact(candidate, config).map((e) => {
+      return {
+        ...e,
+        identifiedType: wasmEntitiesToString(e.identifiedType),
+      };
+    });
+  } else {
+    throw new Error(
+      "redact failed to run because wasm is not supported in this enviornment",
+    );
+  }
+}
+
+type Unredact = (input: string) => string;
+
+export async function redact<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
```

Comment:
Pretty sure this needs another type variable for the redact function.

Reply:
I added one, though it then requires an `as` when we call replace because the Wasm custom entity type is typed as `string`, so the type can't be informed the whole way though.

Reply:
Constant string returns would be a subtype of `string` so it should still work. Would you like me to work on the types?

Reply:
The issue I was having is [here](https://github.com/arcjet/arcjet-js/pull/1358/files#diff-2e25abfb099bd46d41b989f26d6fb2fa881c7a9fde1c97d2af801df3ff1d0f90R134) where we need to pass a `string` in to a function with a more precise type. We know from our own understanding of the code that it should always be valid, but I couldn't find a way of expressing it in TypeScript because we just get a `string` in the generated bindings. If you know a way of making that work properly that would be great.

Reply:
In exploring the type inference on this, I found that the sensitiveInfo inference doesn't work correctly either. We need to fix them.

Reply:
The sample that I've been playing with is:

```ts
const [redacted, unredact] = await redact("abc, 123", {
  // entities: ["credit-card-number", "abc"],
  // detect: (tokens: string[]) => {
  //   return tokens.map((token) => {
  //     if (token == "abc") {
  //       return "abc";
  //     }
  //   });
  // },
  replace: (entity) => {
    // if (entity === "abc") return "123456";
    if (entity === "email") return "foobarbaz";
    // if (entity === "") return "123";
  },
});
```

Essentially, we should be able to comment/uncomment the various pieces and the type inference should be good. I have inlay hints enabled for all my TS so I can see how it is inferring things.

I'm going to push up my changes, but be aware that they have the same caveat as https://github.com/arcjet/arcjet-js/pull/1446#issuecomment-2319084446 and I'm not good enough at TypeScript to know how to make both ways work.